ntoskrnl/ke/i386/cpu.c

```diff
@@ -1359,62 +1425,392 @@ KiCoprocessorError(VOID)
+    ASSERT(FxSaveAreaFrame);
+
+    /* Is NPX state for current thread loaded? */
+    if (FsContext->CurrentThread->NpxState == NPX_STATE_LOADED)
+    {
+        /*
+         * It is, have we gotten the acknowledgement
+         * that the floating point data is allocated
+         * in pool memory?
+         */
+        if (FsContext->PoolAllocated)
+        {
+           /* It is, save the FPU state */
+           Ke386SaveFpuState(FxSaveArea);
+
+           /* And save the NPX CR0 state as well */
+           FxSaveArea->Cr0NpxState = FxSaveAreaFrame->Cr0NpxState;
+           FxSaveArea->NpxSavedCpu = 0;
+        }
+        else
+        {
+            /* It's not, save the control word then */
+            Ke386SaveCwState(FsControlWord);
+
+            /* Get MXCsr if XMMI is supported */
+            if (KeI386XMMIPresent && KeI386FxsrPresent)
+            {
+                FsMXCsr = _mm_getcsr();
+            }
+
+            /* Take away CR0_TS from NPX frame */
+            FxSaveAreaFrame->Cr0NpxState &= ~CR0_TS;
+            FsContext->Cr0NpxState = FxSaveAreaFrame->Cr0NpxState;
+        }
+    }
+    else
+    {
+        /*
+         * NPX state is not loaded, is the
+         * floating point allocated in pool?
+         */
+        if (FsContext->PoolAllocated)
+        {
+            /* It is, copy the NPX frame */
+            RtlCopyMemory(FxSaveArea, FxSaveAreaFrame, sizeof(FX_SAVE_AREA));
+            FxSaveArea->NpxSavedCpu = 0;
+        }
+        else
+        {
+            /*
+             * It's not, save the control word and MXCsr
+             * from frame. First check if FX is supported
+             * and if so then grab those from that.
+             */
+            if (KeI386FxsrPresent)
+            {
+                FsControlWord = FxSaveAreaFrame->U.FxArea.ControlWord;
+                FsMXCsr = FxSaveAreaFrame->U.FxArea.MXCsr;
+            }
+            else
+            {
+                /*
+                 * Otherwise just grab the control word
+                 * from the regular floating point format.
+                 */
+                FsControlWord = FxSaveAreaFrame->U.FnArea.ControlWord;
+            }
+
+            /* Take away CR0_TS from NPX frame */
+            FxSaveAreaFrame->Cr0NpxState &= ~CR0_TS;
+            FsContext->Cr0NpxState = FxSaveAreaFrame->Cr0NpxState;
+        }
+    }
+
+    /* We're almost done, set the stuff we got to the save context */
+    FsContext->MXCsr = FsMXCsr;
+    FsContext->ControlWord = FsControlWord;
+    FsContext->FxSaveArea = FxSaveArea;
+
+    /* Set the current NPX IRQL and store the NPX state */
+    FsContext->CurrentThread->Header.NpxIrql = KeGetCurrentIrql();
+    FsContext->NpxState = FsContext->CurrentThread->NpxState;
+
+    /* Zero out the NPX CR0 state of the frame */
+    FxSaveAreaFrame->Cr0NpxState = 0;
+
+    /* Initialize the FPU and set the default FP control word */
+    Ke386FnInit();
+    Ke386SetCwState(0x27F);
```

Comment:
Naive question: are these values (here and the one below, and maybe elsewhere) flags? (If so it would be of interest to convert them into OR'ed flag values with human-readable names.)

Reply:
Nope, these are just the "raw" (or whatever term it fits for this context of the post) values that are used also when initializing the context of a thread -- see e.g. -> https://git.reactos.org/?p=reactos.git;a=blob;f=ntoskrnl/ke/i386/thrdini.c;hb=fdcefaa7cbd419aef8796b2f690c90d1cd968431#l133

Reply:
quite interesting.... I wonder whether they have a special meaning (flags?) or it's really just, as you call them, "raw", like when the FPU sees this value (and not another one) he will do something.